libswift/Sources/Optimizer/PassManager/PassUtils.swift

```diff
@@ -0,0 +1,115 @@
+//===--- PassUtils.swift - Utilities for optimzation passes ---------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+import OptimizerBridging
+
+public typealias BridgedFunctionPassCtxt =
+  OptimizerBridging.BridgedFunctionPassCtxt
+public typealias BridgedInstructionPassCtxt =
+  OptimizerBridging.BridgedInstructionPassCtxt
+
+struct FunctionPassContext {
+
+  fileprivate let passContext: BridgedPassContext
+  fileprivate let function: Function
+  
+  func erase(instruction: Instruction) {
+    PassContext_eraseInstruction(passContext, instruction.bridged)
+  }
+  
+  private func notifyChanges(_ kind: ChangeNotificationKind) {
+    PassContext_notifyChanges(passContext, kind)
+  }
+}
+
+struct FunctionPass {
+
+  let name: String
+  let runFunction: (Function, FunctionPassContext) -> ()
```

Comment:
Given that this, and `InstructionPass`, just have a name + closure (which I assume will be unique for each pass), would it make sense for it to be a protocol?

Reply:
A pass should not have any state, because the pass object lives forever and it's dangerous to carry over state (like analysis caches) from one run to another. We have this in the C++ optimizer, and it ends up that the run functions of passes always instantiate another "pass" object which does the actual work - a lot of boiler plate. 

Providing the run function as a function, makes this clear.

Reply:
You could also make the run function a static requirement, to enforce that there is no instance and hence no state.

The real benefit of using a protocol is that it gives you the full power of subtyping. So you could create a `protocol SpecialFunctionPass: FunctionPass` with additional requirements, and which provides additional utility functions, etc.

Reply:
Well, in the existing optimizer we didn't have a use for different kind of function passes. I'm not a big fan of adding abstractions/features if there is not a good use for them. I think simplicity should come first - and I think that not needing to define a separate nominal type for a pass makes it a bit simpler.